cpp/src/arrow/extension/variable_shape_tensor.cc

```diff
@@ -0,0 +1,263 @@
+    for (auto v : uniform_shape_) {
+      if (v.has_value()) {
+        uniform_shape.PushBack(v.value(), allocator);
+      } else {
+        uniform_shape.PushBack(rj::Value{}.SetNull(), allocator);
+      }
+    }
+    document.AddMember(rj::Value("uniform_shape", allocator), uniform_shape, allocator);
+  }
+
+  rj::StringBuffer buffer;
+  rj::Writer<rj::StringBuffer> writer(buffer);
+  document.Accept(writer);
+  return buffer.GetString();
+}
+
+Result<std::shared_ptr<DataType>> VariableShapeTensorType::Deserialize(
+    std::shared_ptr<DataType> storage_type, const std::string& serialized_data) const {
+  if (storage_type->id() != Type::STRUCT) {
+    return Status::Invalid("Expected Struct storage type, got ",
+                           storage_type->ToString());
+  }
+  auto value_type = storage_type->field(1)->type()->field(0)->type();
+  const size_t ndim =
+      std::static_pointer_cast<FixedSizeListType>(storage_type->field(0)->type())
+          ->list_size();
+
+  rj::Document document;
+  if (document.Parse(serialized_data.data(), serialized_data.length()).HasParseError()) {
+    return Status::Invalid("Invalid serialized JSON data: ", serialized_data);
+  }
+
+  std::vector<int64_t> permutation;
+  if (document.HasMember("permutation")) {
+    for (auto& x : document["permutation"].GetArray()) {
+      permutation.emplace_back(x.GetInt64());
+    }
+    if (permutation.size() != ndim) {
+      return Status::Invalid("Invalid permutation");
+    }
+  }
+  std::vector<std::string> dim_names;
+  if (document.HasMember("dim_names")) {
+    for (auto& x : document["dim_names"].GetArray()) {
+      dim_names.emplace_back(x.GetString());
+    }
+    if (dim_names.size() != ndim) {
+      return Status::Invalid("Invalid dim_names");
+    }
+  }
+
+  std::vector<std::optional<int64_t>> uniform_shape;
+  if (document.HasMember("uniform_shape")) {
+    for (auto& x : document["uniform_shape"].GetArray()) {
+      if (x.IsNull()) {
+        uniform_shape.emplace_back(std::nullopt);
+      } else {
+        uniform_shape.emplace_back(x.GetInt64());
+      }
+    }
+    if (uniform_shape.size() > ndim) {
+      return Status::Invalid("Invalid uniform_shape");
+    }
+  }
+
+  return variable_shape_tensor(value_type, static_cast<uint32_t>(ndim), permutation,
+                               dim_names, uniform_shape);
+}
+
+std::shared_ptr<Array> VariableShapeTensorType::MakeArray(
+    std::shared_ptr<ArrayData> data) const {
+  DCHECK_EQ(data->type->id(), Type::EXTENSION);
+  DCHECK_EQ("arrow.variable_shape_tensor",
+            static_cast<const ExtensionType&>(*data->type).extension_name());
+  return std::make_shared<ExtensionArray>(data);
+}
+
+Result<std::shared_ptr<Tensor>> VariableShapeTensorType::GetTensor(
+    const std::shared_ptr<ExtensionScalar>& scalar) const {
+  const auto tensor_scalar = internal::checked_pointer_cast<StructScalar>(scalar->value);
+  const auto fields = this->storage_type()->fields();
```

Comment:
Can take const references here
```suggestion
  const auto& tensor_scalar = internal::checked_cast<const StructScalar&>(*scalar->value);
  const auto& fields = this->storage_type()->fields();
```